plugins/core/hu.bme.mit.gamma.expression.model/src/hu/bme/mit/gamma/expression/util/ExpressionModelValidator.java

```diff
@@ -583,63 +595,60 @@ else if (expression instanceof MultiaryExpression) {
+				Declaration declaration = (Declaration) elem;
 				for (VariableDeclaration variableDeclaration : expressionUtil.getReferredVariables(initialExpression)) {
 					if (container == variableDeclaration.eContainer()) {
-						List<EObject> eContents = container.eContents();
+						final EList<EObject> eContents = container.eContents();
 						int elemIndex = eContents.indexOf(elem);
 						int variableIndex = eContents.indexOf(variableDeclaration);
 						if (variableIndex >= elemIndex) {
 							validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
 									"The declarations referenced in the initial value must be declared before the variable declaration.", 
 									new ReferenceInfo(ExpressionModelPackage.Literals.INITIALIZABLE_ELEMENT__EXPRESSION, null)));
+							
 							return validationResultMessages;
 						}
 					}
 				}
 				// Initial value is correct
 				Type variableDeclarationType = declaration.getType();
-				ExpressionType initialExpressionType = typeDeterminator.getType(elem.getExpression());
-				if (!typeDeterminator.equals(variableDeclarationType, initialExpressionType)) {
-					
+				Type initialExpressionType = typeDeterminator2.getType(elem.getExpression());
+				if (!typeDeterminator2.equals(variableDeclarationType, initialExpressionType)) {
 					validationResultMessages.add(new ValidationResultMessage(ValidationResult.ERROR,
 							"The types of the declaration and the right hand side expression are not the same: " +
-											typeDeterminator.transform(variableDeclarationType).toString().toLowerCase() + " and " +
-											initialExpressionType.toString().toLowerCase() + ".", 
+											typePrinter.print(variableDeclarationType) + " and " +
+											typePrinter.print(initialExpressionType) + ".", 
 							new ReferenceInfo(ExpressionModelPackage.Literals.INITIALIZABLE_ELEMENT__EXPRESSION, null)));
-				} 
+				}
 				// Additional checks for enumerations
-				checkEnumerationConformance(variableDeclarationType, initialExpression, ExpressionModelPackage.Literals.INITIALIZABLE_ELEMENT__EXPRESSION);
+				validationResultMessages.addAll(checkEnumerationConformance(variableDeclarationType, initialExpression, ExpressionModelPackage.Literals.INITIALIZABLE_ELEMENT__EXPRESSION));
 				// Additional checks for arrays
 				ArrayTypeDefinition arrayType = null;
-				if (variableDeclarationType instanceof ArrayTypeDefinition) {
-					arrayType = (ArrayTypeDefinition) variableDeclarationType;
+				if (declaration.getType() instanceof ArrayTypeDefinition) {
+					arrayType = (ArrayTypeDefinition) declaration.getType();
 				}
-				else if (variableDeclarationType instanceof TypeReference &&
-						((TypeReference) variableDeclarationType).getReference().getType() instanceof ArrayTypeDefinition) {
-					arrayType = (ArrayTypeDefinition) ((TypeReference) variableDeclarationType).getReference().getType();
+				else if (declaration.getType() instanceof TypeReference &&
```

Comment:
Are you sure this is good? Couldn't this be simplified using an utility method? In Line 628 for arrays, only `declaration.getType()` is used...